Xext/sync.c

```diff
@@ -1284,66 +1284,40 @@ ProcSyncInitialize(ClientPtr client)
 static int
 ProcSyncListSystemCounters(ClientPtr client)
 {
-    xSyncListSystemCountersReply rep = {
-        .type = X_Reply,
-        .sequenceNumber = client->sequence,
-    };
     SysCounterInfo *psci;
-    int len = 0;
-    xSyncSystemCounter *list = NULL, *walklist = NULL;
 
     REQUEST_SIZE_MATCH(xSyncListSystemCountersReq);
 
+    x_rpcbuf_t rpcbuf = { .swapped = client->swapped, .err_clear = TRUE };
+
+    CARD32 nCounters = 0;
     xorg_list_for_each_entry(psci, &SysCounterList, entry) {
-        /* pad to 4 byte boundary */
-        len += pad_to_int32(sz_xSyncSystemCounter + strlen(psci->name));
-        ++rep.nCounters;
-    }
+        CARD16 namelen = strlen(psci->name);
 
-    if (len) {
-        walklist = list = calloc(1, len);
-        if (!list)
-            return BadAlloc;
+        /* write xSyncSystemCounter:
+           the name chars (`namelen` amount of bytes) are directly written
+           after the header fields, then the whole thing is padded to
+           full protocol units.
+        */
+        x_rpcbuf_write_CARD32(&rpcbuf, psci->pCounter->sync.id);
+        x_rpcbuf_write_INT32(&rpcbuf, psci->resolution >> 32);
+        x_rpcbuf_write_INT32(&rpcbuf, psci->resolution);
+        x_rpcbuf_write_CARD16(&rpcbuf, namelen);
+        x_rpcbuf_write_CARD8s(&rpcbuf, (CARD8*)psci->name, namelen);
+        x_rpcbuf_pad(&rpcbuf);
+
+        nCounters++;
     }
 
-    rep.length = bytes_to_int32(len);
+    xSyncListSystemCountersReply rep = {
```

Comment:
We should check for memory allocation (rpcbuf) error before moving forward.

Reply:
yes, of course. thanks for spotting this.

In the longer run, we should think about whether X_SEND_REPLY_WITH_RPCBUF() should check this and return the proper return code. So the final code, instead of:

```
        X_SEND_REPLY_WITH_RPCBUF(client, rep, rpcbuf);
        return Success;
```
would be:

```
        return  X_SEND_REPLY_WITH_RPCBUF(client, rep, rpcbuf);
```